src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceWriter.cs

```diff
@@ -633,9 +633,10 @@ private void WriteGreenFactory(Node nd, bool withSyntaxFactoryContext = false)
             {
                 WriteLine("switch (kind)");
                 OpenBlock();
-                foreach (var kind in nd.Kinds)
+                var kinds = nd.Kinds.Distinct().ToList();
```

Comment:
I didn't follow why we need to add `Distinct()` call in these different locations. What prompted that change?

Reply:
I made it so that if the syntax.xml has N different first and last tokens that they are paired up in the .g4 grammar. Makes the rules actually fit reality. 

However it meant we now emitted duplicate cases when the same token was repeated multiple times in the grammar because it could pair with different things.